pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,279 @@
+		if !ok {
+			// Do stuff to enable scheduling
+			cronSchedule := getCronSchedule(cluster)
+			clusterInfo = &scheduleInfo{
+				cronSchedule: cronSchedule,
+			}
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan enabled for cluster %v: %v",
+				cluster.Name, cronSchedule)
+			if err := scheduleScan(cluster, ssh.clusterScanClient, clusterInfo); err != nil {
+				return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error scheduling scan: %v", err)
+			}
+			ssh.clustersMap[cluster.Name] = clusterInfo
+		} else {
+			if cluster.Spec.ScheduledClusterScan.ScheduleConfig != nil &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != "" &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != clusterInfo.cronSchedule {
+				clusterInfo.cronSchedule = cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule
+				logrus.Infof("scheduledScanHandler: clusterSync: cron schedule modified for cluster %v: %v",
+					cluster.Name, clusterInfo.cronSchedule)
+			}
+		}
+	} else {
+		if ok {
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan disabled for cluster: %v",
+				cluster.Name)
+			delete(ssh.clustersMap, cluster.Name)
+		}
+	}
+	return nil, nil
+}
+
+func (ssh *scheduledScanHandler) sync(key string, _ *v3.ClusterScan) (runtime.Object, error) {
+	if !strings.HasPrefix(key, ScheduledScanKeyPrefix) {
+		return nil, nil
+	}
+	splits := strings.Split(strings.TrimPrefix(key, ScheduledScanKeyPrefix), separator)
+	clusterID := splits[0]
+	pastCronSchedule := splits[1]
+	ssh.Lock()
+	defer ssh.Unlock()
+	clusterInfo, ok := ssh.clustersMap[clusterID]
+	if !ok {
+		logrus.Infof("scheduledScanHandler: sync: scheduled scan has been disabled, ignoring")
+	}
+	cluster, err := ssh.clusterLister.Get("", clusterID)
+	if err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: sync: error fetching cluster %v: %v", clusterID, err)
+	}
+
+	if err := cis.ValidateClusterBeforeLaunchingScan(cluster); err != nil {
+		return nil, err
+	}
+
+	if !cluster.Spec.ScheduledClusterScan.Enabled {
+		logrus.Infof("scheduledScanHandler: sync: scheduled scan is disabled, ignoring the current one")
+		return nil, nil
+	}
+	currentCronSchedule := getCronSchedule(cluster)
+	if currentCronSchedule != pastCronSchedule {
+		logrus.Infof("scheduledScanHandler: sync: cron schedule has changed for cluster: %v, hence skipping",
+			clusterID)
+		return nil, nil
+	}
+
+	if _, err = checkAndLaunchScan(cluster, ssh.clusterClient, ssh.clusterScanClient, clusterInfo); err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: clusterSync: cluster %v: error launching scan: %v",
+			cluster.Name, err)
+	}
+
+	retention := getRetention(cluster)
+	clusterScanClient := ssh.managementClient.ClusterScans(clusterID)
+	if err := cleanOldScans(clusterID, retention, clusterScanClient, ssh.clusterScanLister); err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error cleaning old scans: %v", err)
```

Comment:
I am not sure we should return the error you get from <cleanOldScans> - this error is going to trigger another sync for the same key for which a scan has been already launched

Reply:
This shouldn't be a problem because the checkandlaunchscan is idempotent. If we don't return an error, the clean up of the old scans would be retried until the next scheduled scan. The default is 24 hours, so no clean up for 24 hours.